resource-aggregate/events/servicesMetadataSnapshotTaken.go

```diff
@@ -0,0 +1,328 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	eventTypeServicesMetadataSnapshotTaken = "servicesmetadatasnapshottaken"
+	writeCostAgainstRead                   = 10
+)
+
+func (d *ServicesMetadataSnapshotTaken) Version() uint64 {
+	return d.GetEventMetadata().GetVersion()
+}
+
+func (d *ServicesMetadataSnapshotTaken) Marshal() ([]byte, error) {
+	return proto.Marshal(d)
+}
+
+func (d *ServicesMetadataSnapshotTaken) Unmarshal(b []byte) error {
+	return proto.Unmarshal(b, d)
+}
+
+func (d *ServicesMetadataSnapshotTaken) EventType() string {
+	return eventTypeServicesMetadataSnapshotTaken
+}
+
+func (d *ServicesMetadataSnapshotTaken) AggregateID() string {
+	return commands.MakeServicesResourceUUID(d.GetEventMetadata().GetHubId()).String()
+}
+
+func (d *ServicesMetadataSnapshotTaken) GroupID() string {
+	return d.GetEventMetadata().GetHubId()
+}
+
+func (d *ServicesMetadataSnapshotTaken) IsSnapshot() bool {
+	return true
+}
+
+func (d *ServicesMetadataSnapshotTaken) Timestamp() time.Time {
+	return pkgTime.Unix(0, d.GetEventMetadata().GetTimestamp())
+}
+
+func (d *ServicesMetadataSnapshotTaken) ETag() *eventstore.ETagData {
+	return nil
+}
+
+func (d *ServicesMetadataSnapshotTaken) ServiceID() (string, bool) {
+	return "", false
+}
+
+func (d *ServicesMetadataSnapshotTaken) CopyData(event *ServicesMetadataSnapshotTaken) {
+	d.ServicesMetadataUpdated = event.GetServicesMetadataUpdated()
+	d.EventMetadata = event.GetEventMetadata()
+}
+
+func (d *ServicesMetadataSnapshotTaken) CheckInitialized() bool {
+	return d.GetServicesMetadataUpdated() != nil &&
+		d.GetEventMetadata() != nil
+}
+
+func (d *ServicesMetadataSnapshotTaken) HandleServicesMetadataUpdated(_ context.Context, upd *ServicesMetadataUpdated) (bool, error) {
+	if d.ServicesMetadataUpdated.Equal(upd) {
+		return false, nil
+	}
+	online := make(map[string]*ServicesStatus_Status, len(d.GetServicesMetadataUpdated().GetStatus().GetOnline())+len(upd.GetStatus().GetOnline()))
+	for _, v := range d.GetServicesMetadataUpdated().GetStatus().GetOnline() {
+		online[v.GetId()] = v
+	}
+	offline := make(map[string]*ServicesStatus_Status, len(d.GetServicesMetadataUpdated().GetStatus().GetOffline())+len(upd.GetStatus().GetOffline()))
+	for _, v := range d.GetServicesMetadataUpdated().GetStatus().GetOffline() {
+		offline[v.GetId()] = v
+	}
+	// update current state
+	for _, v := range upd.GetStatus().GetOnline() {
+		online[v.GetId()] = v
+		delete(offline, v.GetId())
+	}
+	for _, v := range upd.GetStatus().GetOffline() {
+		offline[v.GetId()] = v
+		delete(online, v.GetId())
+	}
+	// check if there is no service which is online and offline at the same time
+	for key := range online {
+		if _, ok := offline[key]; ok {
+			return false, fmt.Errorf("invalid status: service %v is online and offline", key)
+		}
+	}
+	for key := range offline {
```

Comment:
Don't need this second iteration, if a key is present in both `online` and `offline` maps then the first iteration will find it.

Reply:
removed